File: pset6/sentimental-credit/credit.py
import cs50
AMEX_LENGTH = 15
MASTERCARD_LENGTH = 16
VISA_LENGTH = 13  # Or 16, but MASTERCARD_LENGTH will check for that already


def main():
    cc_number = cs50.get_string("Number: ")

    # Check if credit card satisfies Luhn's Algorithm
    if luhn_valid(cc_number) == True:
        print(card_type(cc_number))
    else:
        print("INVALID")


# Luhn's Algorithm
def luhn_valid(cc_number):
    sum = 0
    for i in range(len(cc_number) - 2, -1, -2):
        temp = int(cc_number[i]) * 2
        if temp >= 10:
            sum += (temp % 10 + temp // 10)
        else:
            sum += temp
    for i in range(len(cc_number) - 1, -1, -2):
        temp = int(cc_number[i])
        sum += temp
    if sum % 10 == 0:
        return True
    else:
        return False


def card_type(cc_number):
    # Check credit card type
    if len(cc_number) == AMEX_LENGTH and (cc_number[0:2] == "34" or cc_number[0:2] == "37"):
        card_type = "AMEX"
    elif len(cc_number) == MASTERCARD_LENGTH and int(cc_number[0:2]) >= 51 and int(cc_number[0:2]) <= 55:
        card_type = "MASTERCARD"
    elif (len(cc_number) == MASTERCARD_LENGTH or len(cc_number) == VISA_LENGTH) and cc_number[0] == "4":
        card_type = "VISA"
    else:
        card_type = "INVALID"
    return card_type


if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

bool luhn(long n, int k);

int main(void)
{
    const int AMEX_LEN = 15;
    const int MASTERCARD_LEN = 16;
    const int VISA_LEN = 13;

    // Prompt user for credit card info
    long cc_number = get_long("Number: ");

    // Check length of cc
    long temp = cc_number;
    int counter = 0;
    while (temp > 0)
    {
        counter++;
        temp = temp / 10;
    }
    temp = cc_number;

    // If cc is luhn valid, check CC type
    if (luhn(cc_number, counter) == true)
    {
        // Extract first 2 digits of cc
        temp = temp / (pow(10, (counter - 2)));

        // Check card type
        if ((counter == AMEX_LEN) && (temp == 34 || temp == 37))
        {
            printf("AMEX\n");
        }
        else if ((counter == MASTERCARD_LEN) &&
                 (temp == 51 || temp == 52 || temp == 53 || temp == 54 || temp == 55))
        {
            printf("MASTERCARD\n");
        }
        else if ((counter == MASTERCARD_LEN) && (temp / 10 == 4))
        {
            printf("VISA\n");
        }
        else if ((counter == VISA_LEN) && (temp / 10 == 4))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

// Luhn's algorithm
bool luhn(long n, int k)
{
    int sum = 0;
    for (int i = 0; i < k; i++)
    {
        if (i % 2 == 0)
        {
            sum = sum + (n % 10);
        }
        else
        {
            sum = sum + ((2 * (n % 10)) % 10) + ((2 * (n % 10)) / 10);
        }
        n = n / 10;
    }
    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}